### DISCUSSION TITLE
Cannot run insiders with the optimization plugin - Config value "plugins": The "optimize" plugin is not installed

### DISCUSSION THREAD

**Discussion by AnastasiaTWW** [squidfunk/mkdocs-material#6446] [Q&A] [answered]
Hello! I've faced with the following problem: I've installed mkdocs-material 9.4.14+insiders.4.46.0 and try to run the optimize plugin  (including image optimization).

The mkdocs.yml part with the optimize plugin activation looks like this:

```
plugins:
  - search
  - minify:
      minify_html: true
  - meta-descriptions
  - glightbox:
      skip_classes:
        - non-zoomable
  - optimize
```

requirements.txt looks like this:

```
mkdocs==1.5.3
pymdown-extensions==10.5
pygments==2.17.2
mkdocs-minify-plugin==0.7.1
prependnewline==1.2
mkdocs-meta-descriptions-plugin==3.0.0
markdown==3.5.1
importlib_metadata==7.0.0
mkdocs-glightbox==0.3.5
mkdocs-material[imaging]
git+https://${CLONE_INSIDERS_TOKEN}:x-oauth-basic@github.com/<private_fork_of_insiders>.git
```

I'm running the docs build on netlify. The newtlify config file looks like this:

```
[build]
publish = "site"
command = """
pip3 install --no-cache-dir -r requirements.txt &&
mkdocs build -f mkdocs-4.8.yml &&
....
mkdocs build -f mkdocs-pt-BR-4.8.yml
"""
[[headers]]
  for = "/*"
  [headers.values]
    X-Frame-Options = "DENY"
    X-Content-Type-Options = "nosniff"
    Strict-Transport-Security = "max-age=63072000; includeSubDomains"
```

It looks like Netlify installs evertyhing but returns the following error in the end:

```
2:43:52 AM: ERROR   -  Config value "plugins": The "optimize" plugin is not installed
```

Did someone face with the same problem? What am I doing wrong?

**Comment by alexvoss** [comment]
I note that you have `mkdocs-material[imaging]` in your requirements before the line that is meant to install the insiders edition. I am no wizard with pip but if it installs things one-by-one then it will install the public `mkdocs-material` before coming to Insiders. Have a look at your logs, you may find that it says that the dependency `mkdocs-material` is already fulfilled. The solution, then, would be to switch the order of the two lines. Hope this helps.

**Reply by AnastasiaTWW** [reply]
Thank for your reply! I tried to re-order things but it did not help :( The same error

**Reply by alexvoss** [reply]
Can you get a log file that shows what actually happens? It could be that your token is not set or is invalid. Whatever it is, it would seem that you end up running the public version of Material.

**Comment by kamilkrzyskow** [comment, marked as answer]
Hello,
using your example `requirements.txt` (changing of course the URL to Insiders) on Windows, `pip` reported with:
```bash
INFO: pip is looking at multiple versions of mkdocs-material[imaging] to determine which version is compatible with other requirements. This could take a while.
Collecting mkdocs-material[imaging] (from -r requirements.txt (line 10))
Obtaining dependency information for mkdocs-material[imaging] from https://files.pythonhosted.org/packages/a0/06/48127f4ba9cea7c0ccb8a63f8db7c4323caca1301ee42490a636215a87a6/mkdocs_material-9.4.13-py3-none-any.whl.metadata
Downloading mkdocs_material-9.4.13-py3-none-any.whl.metadata (16 kB)
Obtaining dependency information for mkdocs-material[imaging] from https://files.pythonhosted.org/packages/85/9a/e8b05e18c51c528dc91837568eac7fa198ecaa89e4480b9361ac6a676dbe/mkdocs_material-9.4.12-py3-none-any.whl.metadata
  ...
```

"This could take a while" is too long for me, so I changed the entry to `mkdocs-material[imaging]==9.4.14` and `pip` reported with:
```bash
ERROR: Cannot install mkdocs-material 9.4.14+insiders.4.46.0 (from git+https://github.com/...) and mkdocs-material[imaging]==9.4.14 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested mkdocs-material 9.4.14+insiders.4.46.0 (from git+https://github.com/...)
    mkdocs-material[imaging] 9.4.14 depends on mkdocs-material 9.4.14 (from https://files.pythonhosted.org/packages/34/31/98fff7fe1401cf8c45127d836cd34b6d512a5e984f2f4dc3cbb17aad8674/mkdocs_material-9.4.14-py3-none-any.whl (from https://pypi.org/simple/mkdocs-material/) (requires-python:>=3.8))

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

Therefore it seems that there is some sort of conflict with those 2 packages 🤔. The Netlify guide for Insiders also instructs to only use the Insiders version: https://www.starfallprojects.co.uk/projects/deploy-host-docs/deploy-mkdocs-material-netlify/

The `imaging` extra consists of 2 packages, and all other dependency module versions can be managed by the theme, therefore we could simplify the `requirements.txt` to:
```
git+https://${CLONE_INSIDERS_TOKEN}:x-oauth-basic@github.com/<private_fork_of_insiders>.git
pillow==10.1.0
cairosvg==2.7.1
mkdocs-minify-plugin==0.7.1
# Don't know what prependnewline is, so I left it
prependnewline==1.2
mkdocs-meta-descriptions-plugin==3.0.0
# Oddly specific requirement of importlib, so I left it
importlib_metadata==7.0.0
mkdocs-glightbox==0.3.5
```

https://github.com/squidfunk/mkdocs-material/blob/560bb9035dae6c56bfeb1675d2cff8813fe86b70/pyproject.toml#L65-L68

This allowed me to successfully install it with pip.

---

As for the other matter of using Netlify and probably a Linux Docker image, I explained some debugging methods for packages in another discussion: https://github.com/mkdocs/mkdocs/discussions/3492

**Reply by kamilkrzyskow** [reply]
Assuming the other discussion is too abstract (too disconnected from this one) as the other related to a local path and not a installed package like here or TL;DR, then in short you could debug the state of the python installation and environment using some simple commands:
- `python -c "import material.plugins.optimize"` or `python3` instead of `python` to check if plainly running Python will allow you to import the installed `optimize` plugin.
- `python -c "import material; from pathlib import Path; print(*list((Path(material.__file__).parent / 'plugins').iterdir()), sep='\n')"`
to check the contents of the installed material plugins directory.

**Reply by AnastasiaTWW** [reply]
Changing the list of deps seemed to help with the initial problem but another problem appeared - if I use the `optimize` plugin without explicitly specified pngquant among dependencies, I get an error that this dep is not found. Once I use `apt-get install pngquant` during the build (as the pngquant pip version is cut - I've googled it), I get the following

```
1:22:58 PM: E: Could not open lock file /var/lib/dpkg/lock-frontend - open (13: Permission denied)
1:22:58 PM: E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), are you root?
```

**Reply by squidfunk** [reply]
If you don't want to (or are not able to) install pngquant, you can disable .png optimization with [`optimize_png`](onfig.optimize_png):

``` yaml
plugins:
  - optimize:
      optimize_png: false
```

This will still leave .jpg optimization if Pillow is installed.

**Reply by AnastasiaTWW** [reply]
I do want to use png optimization :) I've found the root of the problem. Netlify does not allow to run "apt-get install" https://answers.netlify.com/t/apt-get-install-build-dependency/3095/3, so I install pngquant with `npm install pngquant` and it works fine.

**Reply by squidfunk** [reply]
Perfect! Thanks for sharing ☺️

**Reply by AnastasiaTWW** [reply]
Further note :) 

I get the runtime error when using the optimize plugin and the [meta-description](https://pypi.org/project/mkdocs-meta-descriptions-plugin/) plugin together. Images are compressed, then the meta-desc plugin adds descriptions to the pages and the runtime error comes. This problem is relevant for netlify, honestly I do not know if it repeats in other envs or not and also do not know how to fix this (except to stop using one of the plugins as this helps).

```
6:54:10 PM: INFO    -  Optimized media file: images/waf-installation/quickstart/cdn-node-creation-modal.png 79.0 kB
6:54:10 PM: INFO    -  Optimized media file: images/waf-installation/azure/container-copy-domain-name.png 87.4 kB
6:54:10 PM: INFO    -  Optimized media file: images/waf-installation/quickstart/cname-required-modal.png 101.4 kB
6:54:10 PM: INFO    -  Optimized media file: images/waf-installation/yandex-cloud/create-vm.png 116.4 kB
6:54:11 PM: INFO    -  Optimized media file: images/waf-installation/oob/terraform-module-for-traffic-mirrored-by-server.png 101.7 kB
6:54:43 PM: INFO    -  meta-descriptions: Added meta descriptions to 363 of 370 pages, 363 using the first paragraph
6:54:43 PM: Traceback (most recent call last):
6:54:43 PM:   File "/opt/buildhome/python3.8/bin/mkdocs", line 8, in <module>
6:54:43 PM:     sys.exit(cli())
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
6:54:43 PM:     return self.main(*args, **kwargs)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1078, in main
6:54:43 PM:     rv = self.invoke(ctx)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1688, in invoke
6:54:43 PM:     return _process_result(sub_ctx.command.invoke(sub_ctx))
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
6:54:43 PM:     return ctx.invoke(self.callback, **ctx.params)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 783, in invoke
6:54:43 PM:     return __callback(*args, **kwargs)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/__main__.py", line 286, in build_command
6:54:43 PM:     build.build(cfg, dirty=not clean)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/commands/build.py", line 354, in build
6:54:43 PM:     config.plugins.on_post_build(config=config)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/plugins.py", line 542, in on_post_build
6:54:43 PM:     return self.run_event("post_build", config=config)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/plugins.py", line 509, in run_event
6:54:43 PM:     result = method(**kwargs)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/material/plugins/optimize/plugin.py", line 142, in on_post_build
6:54:43 PM:     raise future.exception()
6:54:43 PM:   File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
6:54:43 PM:     result = self.fn(*self.args, **self.kwargs)
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/material/plugins/optimize/plugin.py", line 274, in _optimize_image
6:54:43 PM:     f.write(json.dumps(self.cache, indent = 2))
6:54:43 PM:   File "/usr/lib/python3.8/json/__init__.py", line 234, in dumps
6:54:43 PM:     return cls(
6:54:43 PM:   File "/usr/lib/python3.8/json/encoder.py", line 201, in encode
6:54:43 PM:     chunks = list(chunks)
6:54:43 PM:   File "/usr/lib/python3.8/json/encoder.py", line 431, in _iterencode
6:54:43 PM:     yield from _iterencode_dict(o, _current_indent_level)
6:54:43 PM:   File "/usr/lib/python3.8/json/encoder.py", line 356, in _iterencode_dict
6:54:43 PM:     for key, value in items:
6:54:43 PM: RuntimeError: dictionary changed size during iteration
```

**Reply by squidfunk** [reply]
Thanks for reporting. So as I understand this only happens when both plugins are active? We don't officially support this plugin, but if you can craft a minimal reproduction that we can use to observe this problem outside of netlify, we're happy to investigate.

**Reply by AnastasiaTWW** [reply]
Unfortunately, I cannot reproduce it with the minimum set of documentation pages and even locally on my machine. However, the error is still in place - I cannot define the exact conditions, to be honest (it does not depend only on meta-descriptions, I've tried to delete it and still got the error). The only condtions I can describe is that there many images to be optimized (~150) and that I run builds on Netlify, where there is no possibility to check the diff with the previously run build, so optimization is run every build for every image we have.

**Reply by kamilkrzyskow** [reply]
The excerpt of the error points to a some kind of race-condition being met when using concurrency:
```bash
6:54:43 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/material/plugins/optimize/plugin.py", line 274, in _optimize_image
6:54:43 PM:     f.write(json.dumps(self.cache, indent = 2))
...
6:54:43 PM:   File "/usr/lib/python3.8/json/encoder.py", line 356, in _iterencode_dict
6:54:43 PM:     for key, value in items:
6:54:43 PM: RuntimeError: dictionary changed size during iteration
```

Did you try changing the configuration of the optimize plugin? @AnastasiaTWW 
If not, disable concurrency by setting the option to 1.
https://squidfunk.github.io/mkdocs-material/plugins/optimize/#config.concurrency

Unfortunately, I don't really work with threading. 
Therefore I can only guess that your Netlify workflow runner, dumps the object to json so slowly, 
that another `_optimize_image` call happens and modifies the dictionary, while the first one didn't save the cache file yet 😆 

Another issue could be the Python version 3.8, which could have some kind of "bug" when it comes to concurrency.
Unfortunately, Netlify doesn't support a newer version https://docs.netlify.com/configure-builds/available-software-at-build-time/#languages
Maybe, create some GitHub workflow pipeline, where you'd build the page on GitHub with Python 3.12 and then deploy it to Netlify? 🤷 

---

But putting "user error" or "workflow runner error" aside, I think the Material theme could move saving of the cache, after the thread pool is finished to avoid any kind of race-conditions, as I don't really see that much upside to saving it after each processed image.

**Reply by kamilkrzyskow** [reply]
Also using the other setting https://squidfunk.github.io/mkdocs-material/plugins/optimize/#config.cache to disable cache only prevents the initial load of the saved cache. It doesn't delete or prevent saving the cache file in the first place. So disabling cache won't prevent it from crashing when using concurrency, which is unfortunate.

**Reply by squidfunk** [reply]
Thanks for digging into this. The reason why we write to `manifest.json` after each optimization process is that we want to keep partial results if the build fails midway. Especially for large documentation sites, this would mean you'd have to run through all steps again. Writing the `manifest.json` after each step allows us to re-use optimizations from prior runs.

I agree that there's a race condition. I haven't run into it yet, but we can surely add a lock when writing the file. We've done the same in the social plugin when downloading fonts and it works like a charm, so locking during writing to `manifest.json` should solve the issue while introducing minimal contention, as the actual optimization is expected to consume the most time.

**Reply by squidfunk** [reply]
In any case, please open an issue for this. I'm sure it is a quick fix then.

**Reply by kamilkrzyskow** [reply]
<del> I guess that saving the cache in `on_build_error` or `on_shutdown` isn't viable, because of the threading? 🤔
If @AnastasiaTWW won't open an issue today then, I will do it tomorrow ✌️ </del>

**Reply by squidfunk** [reply]
Hmm, that might in fact be a much better idea! I haven't came to use those hooks, but let me give this a try.

__Edit__: Thinking about it, using `on_shutdown` we'd miss saving the cache after every build when running `mkdocs serve`, so new images would be optimized again and again until CTRL-C, but writing the cache in `on_post_build` and `on_build_error` should be sufficient.

**Reply by squidfunk** [reply]
I've ended up deferring updating of the manifest to `on_post_build` and `on_shutdown`. Explanation:

- `on_post_build`: triggers after every successful build, including when the project is served
- `on_shutdown`: triggers when the build is interrupted, i.e., through CTRL-C

I'd say `on_build_error` should only be considered a temporary situation, which is either exited by fixing the error, or exiting the process. In both cases, the result is saved by the two other triggers.

I've adjusted the optimize plugin in [`92d3d6b`](https://github.com/squidfunk/mkdocs-material-insiders/commit/92d3d6bc73da22b802f43b9facfff247035ad114), as well as the social plugin in [`a1c137b8`](https://github.com/squidfunk/mkdocs-material-insiders/commit/aa1c137b89c3286f494a171877dc675c9774f019), both of which are using manifest files. @AnastasiaTWW could you check the latest `master` of Insiders and report back whether this fixes your problem?

**Reply by squidfunk** [reply]
Oh, and thanks to @kamilkrzyskow for the fresh perspective! Deferring updating of the manifest makes much more sense, especially when it's the same result without the need for further synchronization primitives.

**Reply by kamilkrzyskow** [reply]
I gave an initial 👍, but looking at the commit again, I see that there is only cache saving in the `on_shutdown` event. As said before I'm no expert in threading, but if something happens before `on_post_build` there won't be any `self.pool.shutdown()` or `future.result()`, so I'm not so sure if a rogue late `_optimize_image` call won't crash the save of cache in `on_shutdown`, as I don't see anything else controlling the additional threads 🤔

**Reply by squidfunk** [reply]
I'm not quite sure what you mean. [Cache is saved in `on_post_build`](https://github.com/squidfunk/mkdocs-material-insiders/blob/92d3d6bc73da22b802f43b9facfff247035ad114/src/plugins/optimize/plugin.py#L153-L156) after all images have been optimized, [and in `on_shutdown`](https://github.com/squidfunk/mkdocs-material-insiders/blob/92d3d6bc73da22b802f43b9facfff247035ad114/src/plugins/optimize/plugin.py#L201-L203), to catch partials results in case something went wrong and _any_ error was triggered. In case `mkdocs serve` is used and an error happens that crashes the build, MkDocs will stop the build, but will keep the plugin state, until the error is resolved. The user can either resolve the error, leading to `on_post_build` being executed after a successful build, or CTRL-C, so the manifest file is saved as well.

Unless I'm missing something.

**Reply by kamilkrzyskow** [reply]
@squidfunk you create the concurrent jobs in `on_env`, and those background jobs are detached from the MkDocs event flow.: 
https://github.com/squidfunk/mkdocs-material-insiders/blob/aa1c137b89c3286f494a171877dc675c9774f019/material/plugins/optimize/plugin.py#L119-L124

Assuming someone has a lot of images and not a lot of documentation, it could lead to MkDocs event flow finishing faster than the background jobs, and if you try to save the cache while the background jobs change the size of the manifest dictionary it will lead to the same error as above.

To prevent such case there is https://devdocs.io/python~3.12/library/concurrent.futures#concurrent.futures.Executor.shutdown
which forces the main thread to wait for the background threads to finish, and you're only using this in `on_post_build`, and only if it's not `served`: https://github.com/squidfunk/mkdocs-material-insiders/blob/aa1c137b89c3286f494a171877dc675c9774f019/material/plugins/optimize/plugin.py#L149-L151

What I'm saying is that when a user invokes CTRL-C after `on_env` and before `on_post_build`, the `on_shutdown` will trigger trying to save cache, while background tasks are possibly still running. CTRL-C after a `serve` could also have this issue I think 🤔 

Hope that makes more sense :v:, but I'm just guessing myself. I would need to prepare some kind of dataset to prove this idea.

---

In other words, I think that `self.pool.shutdown()` should be in `on_post_build` without any condition, as the `self.pool` is set in `on_config` again, and it should be in `on_shutdown` as well. Basically before any cache write to avoid race conditions.

**Reply by squidfunk** [reply]
> In other words, I think that self.pool.shutdown() should be in on_post_build without any condition, as the self.pool is set in on_config again, and it should be in on_shutdown as well. Basically before any cache write to avoid race conditions.

The thread pool is initialized in `on_startup` to keep it alive during `mkdocs serve`, not in `on_config`.

But, I think you're right with the race condition. All concurrent jobs have already been reconciled [in the code block before the shutdown logic](https://github.com/squidfunk/mkdocs-material-insiders/blob/aa1c137b89c3286f494a171877dc675c9774f019/material/plugins/optimize/plugin.py#L138-L147), so we know that there are no more pending jobs and we can write the manifest in  `on_post_build`. However, you're right that jobs are not reconciled in `on_shutdown`, so we need to reconcile before writing the manifest, which is why I've moved the shutdown logic into `on_shutdown`. This is also more symmetrical with `on_startup` setting up the pool. Additionally, I've added the `cancel_futures` flag when shutting down the pool, because either we have a successful build that ran through `on_post_build`, or the user CTRL-C'd in which case we don't want to schedule still unscheduled optimization jobs. Hope I got it right this time.

Commit is [`80277e1`](https://github.com/squidfunk/mkdocs-material-insiders/commit/80277e15790b8a67953eeba0c9d0d80b276ad48e).

**Reply by kamilkrzyskow** [reply]
🤦 I don't know how I didn't see that `on_startup` the first time, and just put everything in `on_config` in my head, sorry for that.
The commit looks good, just the newly imported `wait` seems redundant.
Also unfortunately, as always, I simply checked 3.12 without checking previous versions, and it turns out that the `cancel_futures` is available since Python 3.9, so given that MkDocs supports 3.8 perhaps this setting shouldn't be there. Or support different python versions using `**kwargs` and set a dictionary before calling that function or even a simple if/else would do.
Thanks for your time @squidfunk  👍

**Reply by squidfunk** [reply]
Removed the orphaned `wait` already in a subsequent commit ☺️ Regarding `cancel_futures` – good catch! I forgot to check availability in Python 3.8. I really need to start working on tests that we can run across environments but my todo list is already long enough for 10x myself. I've asked Copilot, and it seems [`613d37a`](https://github.com/squidfunk/mkdocs-material-insiders/commit/613d37a1e6d478a6af7f121f8fc069fec895a35d) is sufficient, so we can keep earlier termination at the expense of 3 LOCs.

Thanks for your time as well, awesome input and threading should be much better now! 🚀

**Reply by AnastasiaTWW** [reply]
I've just run several documentation builds and seems that the problem disappeared! Thank you all for your help and responsiveness! I hope that it will work stable but if the error comes again, I will let you know

**Reply by AnastasiaTWW** [reply]
Hi all! Coming back to this thread as again I have the issue with the image optimization plugin. It occurred once I have updated mkdocs insiders up to 4.53.8 version.

```
4:37:03 PM: INFO    -  Building documentation to directory: /opt/build/repo/site/3.6
4:37:08 PM: Traceback (most recent call last):
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/structure/files.py", line 109, in remove
4:37:08 PM:     del self._src_uris[file.src_uri]
4:37:08 PM: KeyError: '.cache/plugin/optimize/images/installation-azure/auto-scaling/ru/load-balancing-guide/add-balancing-rule.png'
4:37:08 PM: During handling of the above exception, another exception occurred:
4:37:08 PM: Traceback (most recent call last):
4:37:08 PM:   File "/opt/buildhome/python3.8/bin/mkdocs", line 8, in <module>
4:37:08 PM:     sys.exit(cli())
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
4:37:08 PM:     return self.main(*args, **kwargs)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1078, in main
4:37:08 PM:     rv = self.invoke(ctx)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1688, in invoke
4:37:08 PM:     return _process_result(sub_ctx.command.invoke(sub_ctx))
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
4:37:08 PM:     return ctx.invoke(self.callback, **ctx.params)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/click/core.py", line 783, in invoke
4:37:08 PM:     return __callback(*args, **kwargs)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/__main__.py", line 284, in build_command
4:37:08 PM:     build.build(cfg, dirty=not clean)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/commands/build.py", line 319, in build
4:37:08 PM:     env = config.plugins.on_env(env, config=config, files=files)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/plugins.py", line 599, in on_env
4:37:08 PM:     return self.run_event('env', env, config=config, files=files)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/plugins.py", line 566, in run_event
4:37:08 PM:     result = method(item, **kwargs)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/material/plugins/optimize/plugin.py", line 131, in on_env
4:37:08 PM:     files.remove(file)
4:37:08 PM:   File "/opt/buildhome/python3.8/lib/python3.8/site-packages/mkdocs/structure/files.py", line 111, in remove
4:37:08 PM:     raise ValueError(f'{file.src_uri!r} not in collection')
4:37:08 PM: ValueError: '.cache/plugin/optimize/images/installation-azure/auto-scaling/ru/load-balancing-guide/add-balancing-rule.png' not in collection
```

Do you have an idea what is wrong? 

Other deps are the following:

git+https://${CLONE_INSIDERS_TOKEN}:x-oauth-basic@github.com/wallarm/mkdocs-material-insiders.git
pymdown-extensions==10.7
pygments==2.17.2
mkdocs-minify-plugin==0.7.1
prependnewline==1.2
mkdocs-meta-descriptions-plugin==3.0.0
markdown==3.5.2
importlib_metadata==7.0.1
mkdocs-glightbox==0.3.7
pillow==10.3.0
cairosvg==2.7.1

**Reply by squidfunk** [reply]
If you could provide a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/), we're happy to look into it!

**Reply by AnastasiaTWW** [reply]
The thing is that, again, I cannot get the same error in the minimal reproduction. And generally this error seems to be floating. BTW, I quess it is again connected with the Pythin version (we use 3.8 and cannot migrate to the newer one right now as this is the limitation of Netlify). We have ~150 images to be optimized and we have several documentation versions (10). Versioning is implemented with our custom script. Images are optimized only during the build of the newest version. In all other versions of the documentation, we use symbolic links to the images from the newest version.

**Reply by squidfunk** [reply]
Hmm, yeah, I understand that it's hard to track down, but if there's nothing we can observe, it's hard for us to fix 😅

**Reply by pkwarren** [reply]
We ran into this problem regularly on `mkdocs-material@9.6.1-insiders-4.53.15`.

Specifying `concurrency: 1` allowed us to work around the problem (ref: https://squidfunk.github.io/mkdocs-material/plugins/optimize/#config.concurrency):
```
  - optimize:
      concurrency: 1
```

**Comment by AnastasiaTWW** [comment]
Thanks everyone for the help! CLosing the issues - the problem has been caused by Netlify limitations

**Reply by squidfunk** [reply]
I've marked @kamilkrzyskow comment as the answer, because the initial problem was concerning dependencies. The problem with the optimize plugin not finding pngquant is a subsequent problem.
